### PR TITLE
perf(http): optimize `ReadableStream`s backed by a resource

### PR DESCRIPTION
**Pull request by littledivy**
83% faster than Deno 1.20.6
22% faster than Node 16.13.1

`main`:
```
wrk -t8 -c20 -d10s http://localhost:8080/sample.txt
Running 10s test @ http://localhost:8080/sample.txt
  8 threads and 20 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    86.79ms    4.41ms 129.53ms   95.15%
    Req/Sec    23.10      7.35    40.00     83.75%
  1856 requests in 10.10s, 9.09GB read
Requests/sec:    183.78
Transfer/sec:      0.90GB
```

This PR:
```
wrk -t8 -c20 -d10s http://localhost:8080/sample.txt
Running 10s test @ http://localhost:8080/sample.txt
  8 threads and 20 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    14.08ms    2.60ms  52.65ms   91.75%
    Req/Sec   142.92     15.35   161.00     56.44%
  11509 requests in 10.10s, 56.24GB read
Requests/sec:   1139.58
Transfer/sec:      5.57GB
```

Node:
```
Running 10s test @ http://localhost:8081/sample.txt
  8 threads and 20 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    18.50ms    3.05ms  74.03ms   96.78%
    Req/Sec   108.82     13.91   121.00     97.03%
  8754 requests in 10.10s, 42.82GB read
Requests/sec:    866.73
Transfer/sec:      4.24GB
```

sample.txt is 5mb

<details>
<summary> Benchmark code </summary>

```js
// Deno
const server = Deno.listen({ port: 8080 });
console.log("File server running on http://localhost:8080/");

for await (const conn of server) {
  handleHttp(conn);
}

async function handleHttp(conn) {
  const httpConn = Deno.serveHttp(conn);
  for await (const requestEvent of httpConn) {
    // Use the request pathname as filepath
    const url = new URL(requestEvent.request.url);
    const filepath = decodeURIComponent(url.pathname);

    const file = await Deno.open("." + filepath, { read: true });
    const response = new Response(file.readable);
    await requestEvent.respondWith(response);
  }
}
```

```js
// Node
const http = require("http");
const fs = require("fs");

http.createServer(function (req, res) {
    const filename = __dirname + req.url;

    const readStream = fs.createReadStream(filename);

    readStream.on("open", function() {
        readStream.pipe(res);
    });

    readStream.on("error", function(err) {
        res.end(err);
    })
}).listen(8081);
```

</details>

It's not as fast as a dedicated `sendFile` (https://github.com/denoland/deno/pull/14278) but definitely better than `main`

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  for await (const requestEvent of httpConn) {
+    // Use the request pathname as filepath
+    const url = new URL(requestEvent.request.url);
+    const filepath = decodeURIComponent(url.pathname);
+
+    const file = await Deno.open("." + filepath, { read: true });
+    // const readableStream = readableStreamFromReader(file);
+    const response = new Response(file.readable);
+    await requestEvent.respondWith(response);
+    Deno.core.tryClose(file.rid);
```
Comment (littledivy) [inline server.js:21]: The op should try_close.

File: serde_v8/magic/buffer.rs [hunk 2]
```diff
@@ -14,19 +14,26 @@ use crate::magic::transl8::impl_magic;
 pub enum MagicBuffer {
   FromV8(ZeroCopyBuf),
   ToV8(Mutex<Option<Box<[u8]>>>),
+  Temp(Vec<u8>),
```
Comment (bartlomieju) [inline serde_v8/magic/buffer.rs:17]: Would be nice to get a comment explaining what this "Temp" means and when it's used

File: ext/web/06_streams.js [hunk 3]
```diff
@@ -633,6 +634,41 @@
     return stream[_disturbed];
   }
 
+  const DEFAULT_CHUNK_SIZE = 16_640;
```
Comment (bartlomieju) [inline ext/web/06_streams.js:637]: Have you experimented with greater size of this buffer?

File: ext/web/06_streams.js [hunk 4]
```diff
@@ -4278,6 +4314,7 @@
     WeakMapPrototypeSet(countSizeFunctionWeakMap, globalObject, size);
   }
 
+  const _maybeRid = Symbol("maybeRid");
```
Comment (aapoalas) [inline ext/web/06_streams.js:4317]: Align with common symbol "language" and use `[[maybeRid]]` perhaps?

File: ext/web/06_streams.js (hunk 3 above)
Comment (littledivy) [reply ext/web/06_streams.js:637]: This change in unrelated to the optimization. It's just moving existing code from `ext/net` to `ext/web/06_streams.js`.  
Though yes, we should try increasing the chunk size to smth like 64KB, once #14265 lands to actually benefit from >16kb sizes.

File: ext/http/lib.rs [hunk 5]
```diff
@@ -664,6 +665,57 @@ async fn op_http_write_headers(
...
+      }
+      HttpResponseWriter::Closed => {
+        return Err(http_error("response already completed"))
+      }
+    };
+
+    let mut vec = vec![0u8; 16 * 1024];
+    let vec_ptr = vec.as_mut_ptr();
+    let buf = ZeroCopyBuf::new_temp(vec);
+    let nread = resource.clone().read(buf).await?;
```
Comment (ry) [inline ext/http/lib.rs:694]: Is the vec stack allocated here? Wouldn't that be a problem across the await boundary?

maybe use [into_boxed_slice](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.into_boxed_slice)?

Comment (AaronO) [reply ext/http/lib.rs:694]: > Is the vec stack allocated here? Wouldn't that be a problem across the await boundary?
> 
> maybe use [into_boxed_slice](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.into_boxed_slice)?

Vecs are always heap allocated, it's wrapped in a ZeroCopyBuf to satisfy the resource read trait sig.

As previously mentioned ZeroCopyBuf is conceptually similar to a Cow enum of Owned/Borrowed, Temp is hackish but I'll clean that up

File: ext/http/lib.rs [hunk 6]
```diff
@@ -664,6 +665,57 @@ async fn op_http_write_headers(
...
+      HttpResponseWriter::Body(body_tx) => body_tx,
+      HttpResponseWriter::Headers(_) => {
+        return Err(http_error("no response headers"))
+      }
+      HttpResponseWriter::Closed => {
+        return Err(http_error("response already completed"))
+      }
+    };
+
+    let mut vec = vec![0u8; 16 * 1024];
```
Comment (ry) [inline ext/http/lib.rs:691]: add TODO to increase this to 64k

File: ext/web/06_streams.js (hunk 3 above)
Comment (littledivy) [reply ext/web/06_streams.js]: updated! and holy moly we are faster than Node now (see updated benchmarks above)

**Review by bartlomieju — APPROVED** [review]
LGTM, great work Divy!

**Review by lucacasonato — APPROVED** [review]
LGTM!

**Review by lucacasonato — CHANGES_REQUESTED** [review]
Actually, a few issues I just realized this has that need to be addressed:

a) before attempting to stream the body using the rid, we need to make sure the stream is not locked
b) we need to lock the stream while we are reading
c) we need to close the stream once we are done reading

If we don't do that last thing, the underlying resource will not be automatically closed by the `ReadableStream` implementation. This also means that `op_http_write_resource` should not close the stream resource itself.

We'll need tests for these cases.

**Comment by lucacasonato** [comment]
Can we get a test for the locking scenarios?

**Review by lucacasonato — APPROVED** [review]
LGTM!

File: ext/http/lib.rs [hunk 7]
```diff
@@ -664,6 +665,56 @@ async fn op_http_write_headers(
...
+    let mut vec = vec![0u8; 64 * 1024];
+    let vec_ptr = vec.as_mut_ptr();
+    let buf = ZeroCopyBuf::new_temp(vec);
+    let nread = resource.clone().read(buf).await?;
+    if nread == 0 {
+      break;
+    }
+    // SAFETY: ZeroCopyBuf keeps the Vec<u8> alive.
+    let bytes =
+      Bytes::from_static(unsafe { std::slice::from_raw_parts(vec_ptr, nread) });
```
Comment (lucacasonato) [inline ext/http/lib.rs:700]: Looking over this code again, this isn't safe. `ZeroCopyBuf` is moved into `resource.read()`, and as such can already be dropped at this point (add a `drop` handler to `ZeroCopyBuf` that prints something and observe).

File: ext/http/01_http.js [hunk 8]
```diff
@@ -269,35 +269,50 @@
...
+            if (respBody.locked) {
+              throw new TypeError("ReadableStream is locked.");
             }
+            const _reader = respBody.getReader(); // Aquire JS lock.
+            await core.opAsync(
+              "op_http_write_resource",
+              streamRid,
+              resourceRid,
+            );
+            readableStreamClose(respBody); // Release JS lock.
```
Comment (dsherret) [inline ext/http/01_http.js:283]: @littledivy if the above statement throws should this still be released?

Comment (littledivy) [reply ext/http/01_http.js:283]: oh nice catch, yeah it should call `reader.cancel`

Comment (dsherret) [reply ext/http/01_http.js:283]: Cool. We can fix it when getting this PR back into main.

Comment (littledivy) [reply ext/http/01_http.js:283]: Fixed in https://github.com/denoland/deno/pull/14345

**Pull request merged**